Assets/_BForBoss/_Utility/Scripts/ObjectPooler.cs

```diff
@@ -0,0 +1,130 @@
+using System;
+using UnityEngine;
+#if UNITY_2021_1_OR_NEWER
+using UnityEngine.Pool;
+#else
+using UnityEngine.SceneManagement;
+using System.Collections.Generic;
+#endif
+
+namespace Perigon.Utility
+{
+    public class ObjectPooler<T> where T : Component
+    {
+		protected string name;
+
+#if UNITY_2021_1_OR_NEWER
+		protected IObjectPool<T> objectPool;
+#else
+		protected Scene poolScene;
+		protected List<T> pool = null;
+#endif
+
+		protected Func<T> buildObject;
+		protected Action<T> actionOnGet;
+		protected Action<T> actionOnRelease;
+
+		/// <summary>
+		/// Create a new Object pooler for a given type
+		/// </summary>
+		/// <param name="poolName">Name of the pool, to be used for the pool scene name</param>
+		/// <param name="buildObject">Function that creates a new instance of an object when existing object is available</param>
+		/// <param name="actionOnGet">Setup action on grabbing a new object from the pool</param>
+		/// <param name="actionOnRelease">Clean up action when recycling an object</param>
+		public ObjectPooler(string poolName, Func<T> buildObject, Action<T> actionOnGet = null, Action<T> actionOnRelease = null)
+		{
+			this.name = poolName;
+			this.buildObject = buildObject;
+			this.actionOnGet = actionOnGet;
+			this.actionOnRelease = actionOnRelease;
+
+#if UNITY_2021_1_OR_NEWER
+			objectPool = new ObjectPool<T>(buildObject, actionOnGet, actionOnRelease);
+#endif
+		}
+
+        ~ObjectPooler()
+        {
+#if UNITY_2021_1_OR_NEWER
+			objectPool.Dispose();
+#else
+			GameObject[] rootObjects = poolScene.GetRootGameObjects();
+			for (int i = 0; i < rootObjects.Length; i++)
+			{
+                UnityEngine.Object.Destroy(rootObjects[i]);
+			}
+			pool = null;
+#endif
+		}
```

Comment:
I'm not sure how necessary this is cause I think it gets done anyways when the scene unloaded, but I'm not really sure